tmt/steps/report/junit.py

```diff
@@ -111,36 +68,112 @@ class ReportJUnitData(tmt.steps.report.ReportStepData):
+
+    @functools.cached_property
+    def duration(self):
+        return sum([duration_to_seconds(r.duration) for r in self._results])
+
+
 @tmt.steps.provides_method('junit')
 class ReportJUnit(tmt.steps.report.ReportPlugin[ReportJUnitData]):
     """
-    Save test results in JUnit format.
+    Save test results in chosen JUnit flavor format. When flavor is set to
+    custom, the ``template_path`` with a path to a custom template must be
+    provided.
 
     When ``file`` is not specified, output is written into a file
     named ``junit.xml`` located in the current workdir.
     """
 
     _data_class = ReportJUnitData
 
+    def check_options(self) -> None:
+        """ Check the module options """
+
+        if self.data.flavor and self.data.template_path:
+            raise tmt.utils.ReportError("The options '--template-path' and "
+                                        "'--flavor' are mutually exclusive.")
+
+        self.data.flavor = self.data.flavor or DEFAULT_FLAVOR_NAME
+
     def prune(self, logger: tmt.log.Logger) -> None:
         """ Do not prune generated junit report """
 
     def go(self, *, logger: Optional[tmt.log.Logger] = None) -> None:
         """ Read executed tests and write junit """
         super().go(logger=logger)
 
-        junit_xml = import_junit_xml()
-        suite = make_junit_xml(self)
+        self.check_options()
 
         assert self.workdir is not None
         f_path = self.data.file or self.workdir / DEFAULT_NAME
+
+        # Get the template according to chosen flavor
+        if self.data.flavor == 'custom':
+            template_path = Path(self.data.template_path)
+        else:
+            template_path = tmt.utils.resource_files(
+                f'steps/report/junit/{self.data.flavor}.xml.j2')
+
+        # Get the results summary
+        results_context = ResultSummaryContext(self.step.plan.execute.results())
+
+        # Prepare the template environment
+        def _read_log(log):
+            try:
+                return self.step.plan.execute.read(log)
+            except AttributeError:
+                return ''
+
+        environment = tmt.utils.default_template_environment()
+        environment.filters.update({
+            'read_log': _read_log,
+            })
+
         try:
-            with open(f_path, 'w') as fw:
-                if hasattr(junit_xml, 'to_xml_report_file'):
-                    junit_xml.to_xml_report_file(fw, [suite])
-                else:
-                    # For older junit-xml
-                    junit_xml.TestSuite.to_file(fw, [suite])
+            self.write(
+                f_path,
+                data=tmt.utils.render_template_file(
+                    template_path,
+                    environment,
+                    results=results_context,
+                    base_dir=self.step.plan.execute.workdir,
+                    plan=self.step.plan,
+                    include_output_log=self.data.include_output_log))
```

Comment:
* I'd recommend using upper-cased names for our custom variables injected into the field, i.e. `RESULTS`, `BASE_DIR`, ... It makes them stand out in the template.
* how about `WORKDIR` instead of `base_dir`? tmt does not have many "base" directories, but "workdir" is a very common thing, sticking to the name would make the story more consistent.